data_safe_haven/config/pulumi.py

```diff
@@ -0,0 +1,147 @@
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Annotated
+
+import yaml
+from pydantic import BaseModel, PlainSerializer, ValidationError
+from pydantic.functional_validators import AfterValidator
+from yaml import YAMLError
+
+from data_safe_haven.config import Context
+from data_safe_haven.exceptions import (
+    DataSafeHavenConfigError,
+    DataSafeHavenParameterError,
+)
+from data_safe_haven.external import AzureApi
+from data_safe_haven.functions import b64decode, b64encode
+from data_safe_haven.utility.annotated_types import UniqueList
+
+
+def base64_string_decode(v: str) -> str:
+    """Pydantic validator function for decoding base64"""
+    return b64decode(v)
+
+
+B64String = Annotated[
+    str,
+    PlainSerializer(b64encode, return_type=str),
+    AfterValidator(base64_string_decode),
+]
+
+
+class PulumiStack(BaseModel, validate_assignment=True):
+    """Container for Pulumi Stack persistent information"""
+
+    name: str
+    config: B64String
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, PulumiStack):
+            return NotImplemented
+        return self.name == other.name or self.config == other.config
+
+    def __hash__(self) -> int:
+        return hash(self.name)
+
+    def write_config(self, context: Context) -> None:
+        """Write stack configuration to a YAML file"""
+        work_dir = Path(context.work_directory / "pulumi" / self.name)
+        if not work_dir.exists():
+            work_dir.mkdir(parents=True)
+        config_path = work_dir / f"Pulumi.{self.name}.yaml"
+        with open(config_path, "w") as f_config:
+            f_config.write(self.config)
+
+
+class PulumiConfig(BaseModel, validate_assignment=True):
+    stacks: UniqueList[PulumiStack]
+
+    def __getitem__(self, key: str) -> PulumiStack:
+        if not isinstance(key, str):
+            msg = "'key' must be a string."
```

Comment:
```suggestion
            msg = "'{key}' must be a string."
```

Reply:
This is intentional, 'key' the argument rather than `{key}` the value.